sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetReadState.java

```diff
@@ -33,31 +51,102 @@
   /** The remaining number of values to read in the current batch */
   int valuesToReadInBatch;
 
-  ParquetReadState(int maxDefinitionLevel) {
+  ParquetReadState(int maxDefinitionLevel, PrimitiveIterator.OfLong rowIndexes) {
     this.maxDefinitionLevel = maxDefinitionLevel;
+    this.rowRanges = rowIndexes == null ? null : constructRanges(rowIndexes);
+    nextRange();
+  }
+
+  private Iterator<RowRange> constructRanges(PrimitiveIterator.OfLong rowIndexes) {
```

Comment:
Can we add some comments to explain how to generate row ranges?

Reply:
Yup will do